chainio/txmgr/txmgr.go

```diff
@@ -2,209 +2,35 @@ package txmgr
-		}
-	}
-}
-
-func (m *SimpleTxManager) queryReceipt(ctx context.Context, txID wallet.TxID) *types.Receipt {
-	receipt, err := m.wallet.GetTransactionReceipt(ctx, txID)
-	if errors.Is(err, ethereum.NotFound) {
-		m.log.Info("Transaction not yet mined", "txID", txID)
-		return nil
-	} else if err != nil {
-		m.log.Info("Receipt retrieval failed", "txID", txID, "err", err)
-		return nil
-	} else if receipt == nil {
-		m.log.Warn("Receipt and error are both nil", "txID", txID)
-		return nil
-	}
-
-	return receipt
-}
-
-// estimateGasAndNonce we are explicitly implementing this because
-// * We want to support legacy transactions (i.e. not dynamic fee)
-// * We want to support gas management, i.e. add buffer to gas limit
-func (m *SimpleTxManager) estimateGasAndNonce(ctx context.Context, tx *types.Transaction) (*types.Transaction, error) {
-	gasTipCap, err := m.client.SuggestGasTipCap(ctx)
-	if err != nil {
-		// If the transaction failed because the backend does not support
-		// eth_maxPriorityFeePerGas, fallback to using the default constant.
-		m.log.Info("eth_maxPriorityFeePerGas is unsupported by current backend, using fallback gasTipCap")
-		gasTipCap = FallbackGasTipCap
-	}
-
-	header, err := m.client.HeaderByNumber(ctx, nil)
-	if err != nil {
-		return nil, err
-	}
-
-	// 2*baseFee + gasTipCap makes sure that the tx remains includeable for 6 consecutive 100% full blocks.
-	// see https://www.blocknative.com/blog/eip-1559-fees
-	gasFeeCap := new(big.Int).Add(header.BaseFee.Mul(header.BaseFee, big.NewInt(2)), gasTipCap)
-
-	gasLimit := tx.Gas()
-	// we only estimate if gasLimit is not already set
-	if gasLimit == 0 {
-		from, err := m.wallet.SenderAddress(ctx)
-		if err != nil {
-			return nil, errors.Join(errors.New("send: failed to get sender address"), err)
-		}
-		gasLimit, err = m.client.EstimateGas(ctx, ethereum.CallMsg{
-			From:      from,
-			To:        tx.To(),
-			GasTipCap: gasTipCap,
-			GasFeeCap: gasFeeCap,
-			Value:     tx.Value(),
-			Data:      tx.Data(),
-		})
-		if err != nil {
-			return nil, errors.Join(errors.New("send: failed to estimate gas"), err)
-		}
-	}
-
-	rawTx := &types.DynamicFeeTx{
-		ChainID:   tx.ChainId(),
-		To:        tx.To(),
-		GasTipCap: gasTipCap,
-		GasFeeCap: gasFeeCap,
-		Data:      tx.Data(),
-		Value:     tx.Value(),
-		Gas:       gasLimit,
-		Nonce:     tx.Nonce(), // We are not doing any nonce management for now but we probably should later for more robustness
-	}
-
-	return types.NewTx(rawTx), nil
+// TxnManager receives transactions from the caller, sends them to the chain, and monitors their status.
+// It also handles the case where a transaction is not mined within a certain time. In this case, it will
+// resend the transaction with a higher gas price. It is assumed that all transactions originate from the
+// same account.
+type AsyncTxManager interface {
+	Start(ctx context.Context)
+	ProcessTransaction(ctx context.Context, req *TxnRequest) error
+	ReceiptChan() chan *ReceiptOrErr
```

Comment:
This I'm not sure what to do with. The eigenda txmgr was implemented using this async interface, whereas our simple txmgr is using a blocking sync interface.

Should we keep them separate like this? Or should we convert one to the other? Def making the simple txmgr into an async version is easier, but is that what we want?

Reply:
resolving because decided to move to sync api